src/node/txreconciliation.cpp

```diff
@@ -376,6 +377,38 @@ class TxReconciliationTracker::Impl
 
         return IsFanoutTarget(wtxid, peer_id, peer_state->m_we_initiate, n);
     }
+
+    std::vector<NodeId> SortPeersByFewestParents(std::vector<Wtxid> parents) EXCLUSIVE_LOCKS_REQUIRED(!m_txreconciliation_mutex)
```

Comment:
fd7c35a22669d440afe182a153bed28fd326b51f

Often there will be no parents (a parent is in the mempool, but been flooded/reconciled already). It would be nice to give a random order in that case.

Reply:
This is only called if there are in mempol parents though, the ordering doesn't matter much otherwise

Reply:
In that case, no transaction would be moved from the reconciliation set to `invs_to_send`, given no transaction would be removed:

```cpp
auto it = std::find(fanout_with_ancestors.begin(), fanout_with_ancestors.end(), peer_id);
if (it != fanout_with_ancestors.end()) {
    for (const auto wtxid: parents) {
        m_txreconciliation->TryRemovingFromSet(peer_id, wtxid);
        invs_to_send.push_back(wtxid);
    }
} else {
    // If the peer is registered for set reconciliation, maybe pick it as fanout
    fanout = m_txreconciliation->ShouldFanoutTo(peer_id, fanout_targets);
}
```

This may be a bit counterintuitive. I'll give some though and potentially re-design so it clearer

Reply:
Say there is an in-mempool parent announced earlier, so it's not in any of the sets.
`SortPeersByFewestParents` will return the same order as `m_states` (since all `parent_count` are 0),
and `fanout_with_ancestors` will take first two after resize.

Then for two peers (specifically, the first two from `m_states`) this code is executed:
```
                for (const auto wtxid: parents) {
                    m_txreconciliation->TryRemovingFromSet(peer_id, wtxid);
                    invs_to_send.push_back(wtxid);
                }
```

`TryRemovingFromSet` will do nothing. But then `invs_to_send.push_back(wtxid);` is executed anyway. And always for the same two first peers from `m_states`. In other words, the first two are guaranteed to take the fanout in this case.

Reply:
Sorry, you're right. I've been looking at this being conditional to `m_txreconciliation->TryRemovingFromSet(peer_id, wtxid`, which is what happens with the descendants, but here it is unconditional.

It should be parsed only if `TryRemovingFromSet` succeeds.

Reply:
Mave the move dependant on `TryRemovingFromSet`